### PR TITLE
New: adds a document denied redirect page setting

### LINKED ISSUE #586: Custom frontend redirect page setting for when document access not allowed
When a user is not allowed to access a document in the frontend, we could provide a new setting to enter a custom redirect page link instead of a white page with a single message.

### PR DESCRIPTION
**Pull request by alexmigf**
closes #586 

This PR adds a new selector for a document access denied redirect page in the frontend:

Screenshot from 2023-08-24 12-28-05

Instead of dying immediately. Other options are:

- Login page
- My Account page
- Custom page

### CODE REVIEW

File: includes/wcpdf-functions.php [hunk 1]
```diff
@@ -341,3 +341,23 @@ function wcpdf_convert_encoding( $string, $tool = 'mb_convert_encoding' ) {
...
+/**
+ * Safe redirect or die.
+ *
+ * @param  string          $url
+ * @param  string|WP_Error $message
+ * @return void
+ */
+function wcpdf_safe_redirect_or_die( $url = '', $message = '' ) {
+	if ( ! empty( $url ) ) {
+		wp_safe_redirect( $url );
```
Comment (MohamadNateqi) [inline includes/wcpdf-functions.php:354]: I was testing to see if it redirected to Google. But then I understood it allows only the domain specified as Site Address (URL) in settings.
Just wanted to ask do you think users may wish to redirect to other domains out of their website.

File: includes/class-wcpdf-main.php [hunk 2]
```diff
@@ -304,11 +308,13 @@ public function generate_pdf_ajax() {
...
 		if ( empty( $_REQUEST['order_ids'] ) ) {
-			wp_die( esc_attr__( "You haven't selected any orders", 'woocommerce-pdf-invoices-packing-slips' ) );
+			$message = esc_attr__( "You haven't selected any orders", 'woocommerce-pdf-invoices-packing-slips' );
+			wcpdf_safe_redirect_or_die( $redirect_url, $message );
 		}
 
 		if ( empty( $_REQUEST['document_type'] ) ) {
-			wp_die( esc_attr__( 'Some of the export parameters are missing.', 'woocommerce-pdf-invoices-packing-slips' ) );
+			$message = esc_attr__( 'Some of the export parameters are missing.', 'woocommerce-pdf-invoices-packing-slips' );
+			wcpdf_safe_redirect_or_die( $redirect_url, $message );
```
Comment (MohamadNateqi) [inline includes/class-wcpdf-main.php:317]: Same here.
And there are a few below that I don't mention to prevent repetition.

**Review by MohamadNateqi — CHANGES_REQUESTED** [review]

File: includes/wcpdf-functions.php (hunk 1 above)
Comment (alexmigf) [reply includes/wcpdf-functions.php:354]: I don't think so.

File: includes/class-wcpdf-main.php [hunk 3]
```diff
@@ -304,11 +308,13 @@ public function generate_pdf_ajax() {
 		}
 
 		if ( empty( $_REQUEST['order_ids'] ) ) {
-			wp_die( esc_attr__( "You haven't selected any orders", 'woocommerce-pdf-invoices-packing-slips' ) );
+			$message = esc_attr__( "You haven't selected any orders", 'woocommerce-pdf-invoices-packing-slips' );
+			wcpdf_safe_redirect_or_die( $redirect_url, $message );
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply includes/class-wcpdf-main.php:312]: You're right, I will fix.

**Review by MohamadNateqi — APPROVED** [review]
Perfect!

**Comment by dwalkerpriv** [comment]
Whatever access type I select with a custom link, it redirects me to the login page.

Screenshot 2023-08-25 120700


When I have access type of 'Full'  or 'Logged in' without a custom link, I can download the PDF. But if i change the permission to 'Guest'  without a custom link, i get the no permission page.

File: includes/class-wcpdf-endpoint.php [hunk 4]
```diff
@@ -177,6 +177,37 @@ public function get_document_link_access_type() {
...
+		$redirect_url   = '';
+		$debug_settings = get_option( 'wpo_wcpdf_settings_debug', array() );
+		
+		if ( isset( $debug_settings['document_access_denied_redirect_page'] ) ) {
+			switch ( $debug_settings['document_access_denied_redirect_page'] ) {
+				case 'login_page':
+					$redirect_url = wp_sanitize_redirect( wp_login_url() );
+					break;
+				case 'myaccount_page':
+					$redirect_url = wp_sanitize_redirect( wp_login_url( wc_get_page_permalink( 'myaccount' ) ) );
```
Comment (YordanSoares) [inline includes/class-wcpdf-endpoint.php:195]: All the options are work correctly, except this one.

I think we need to remove the `wp_login_url()` function here, isn't?

**Review by YordanSoares — CHANGES_REQUESTED** [review]

File: includes/class-wcpdf-endpoint.php (hunk 4 above)
Comment (alexmigf) [reply includes/class-wcpdf-endpoint.php:195]: Are you logged in?

**Comment by alexmigf** [comment]
@dwalkerpriv external links are not allowed, could you try with a store frontend link?

**Comment by alexmigf** [comment]
> Works as intended this time. I didn't notice the description about using frontend pages

Actually I added today after your comment :)

File: includes/class-wcpdf-endpoint.php (hunk 4 above)
Comment (YordanSoares) [reply includes/class-wcpdf-endpoint.php:195]: @alexmigf I tested again in this is the result:

- Logged in: It redirects to the WP dashboard
- Guest user: It redirects to login page, not the My account page.

**Review by YordanSoares — APPROVED** [review]
LGTM! 🚀

**Pull request merged**